consensus/tendermint/tendermint.go

```diff
@@ -0,0 +1,637 @@
+package tendermint
+
+import (
+	"fmt"
+	"maps"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/NethermindEth/juno/core/felt"
+)
+
+type step uint
+
+const (
+	propose step = iota
+	prevote
+	precommit
+)
+
+type timeoutFn func(round uint) time.Duration
+
+type Addr interface {
+	// Ethereum Addresses are 20 bytes
+	~[20]byte | felt.Felt
+}
+
+type Hash interface {
+	~[32]byte | felt.Felt
+}
+
+// Hashable's Hash() is used as ID()
+type Hashable[H Hash] interface {
+	Hash() H
+}
+
+type Application[V Hashable[H], H Hash] interface {
+	// Value returns the value to the Tendermint consensus algorith which can be proposed to other validators.
+	Value() V
+
+	// Valid returns true if the provided value is valid according to the application context.
+	Valid(V) bool
+}
+
+type Blockchain[V Hashable[H], H Hash, A Addr] interface {
+	// Height return the current blockchain height
+	Height() uint
+
+	// Commit is called by Tendermint when a block has been decided on and can be committed to the DB.
+	Commit(uint, V, []Precommit[H, A])
```

Comment:
What happens if Commit() fails / generates an error? Do we just log it / panic?

Reply:
This is something I was hoping to decide later on. However, the idea is that we tell the caller that a consensus has been reached for the current height. After that, it is up to the caller what they want to do with that information.

Then the consensus algorithm optimistically moves to the next height but if it turns out for some reason `Height()` has not been incremented then the caller should decide what to do. The caller will be the synchroniser.